go/mcap/new_reader.go

```diff
@@ -0,0 +1,510 @@
+package mcap
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"math"
+	"sort"
+
+	"github.com/klauspost/compress/zstd"
+	"github.com/pierrec/lz4/v4"
+)
+
+type ReadOptions struct {
+	// optional callback which returns whether to include a given channel of messages in the results.
+	ChannelFilter func(schema *Schema, channel *Channel) bool
+	// optional callback which returns whether to include a given attachment in the results.
+	AttachmentFilter func(name string) bool
+	// optional callback which returns whether to include a given metadata record in the results.
+	MetadataFilter func(name string) bool
+	// Messages and Attachments with LogTimes in the range [StartTime, EndTime) will be included.
+	StartTime uint64
+	EndTime   uint64
```

Comment:
I think that aspirationally anyway, the reader should be designed to be reusable, in that I should be able to instantiate a reader and make multiple queries against it without needing to allocate new buffers. I think there are various complications with this today, but I think this API may add some more. Some considerations -
* Instantiation of the lexer is going to allocate some buffers, which the lexer will grow to size with use.
* Putting StartTime, EndTime, and the topic list  (via channel filter) into the constructor for the reader, rather than in the equivalent to the rosbag bag.Messages call, means that StartTime, EndTime, Topics have the same lifetime as the costly lexer.
* This causes issues for reuse because I need to make a new lexer every time I want to query different data in the file.

Would it make sense to pull StartTime, EndTime, and ChannelFilter (or some manifestation of) out of the ReadOptions, and instead stick those in the call to Read? And also, to attach the lexer to the reader so that it can be reused.